File: dotnet/src/Cekat.EventSdk/Stripe/StripeMetadata.cs
using System.Collections.ObjectModel;

namespace Cekat.EventSdk;

/// <summary>Builds native Stripe metadata for optional Cekat visitor correlation.</summary>
public static class StripeMetadata
{
    private const string VisitorMetadataKey = "cekat_visitor_id";

    /// <summary>Returns fresh read-only metadata for a valid explicit visitor.</summary>
    public static IReadOnlyDictionary<string, string> ForVisitor(string? visitorId)
    {
        var normalizedVisitorId = ValidVisitorId(visitorId);
        return normalizedVisitorId is null
            ? EmptyMetadata()
            : ReadOnlyMetadata(new Dictionary<string, string> { [VisitorMetadataKey] = normalizedVisitorId });
    }

    /// <summary>Returns fresh read-only metadata from the supplied or shared visitor scope.</summary>
    public static IReadOnlyDictionary<string, string> FromCurrentVisitor(IVisitorContext? context = null)
    {
        return ForVisitor((context ?? AsyncLocalVisitorContext.Shared).CurrentVisitorId);
    }

    /// <summary>Copies metadata and replaces only the Cekat visitor key when the visitor is valid.</summary>
    public static IReadOnlyDictionary<string, string> MergeMetadata(IReadOnlyDictionary<string, string> metadata, string? visitorId)
    {
        var merged = new Dictionary<string, string>(metadata);
        var normalizedVisitorId = ValidVisitorId(visitorId);
        if (normalizedVisitorId is not null)
        {
            merged[VisitorMetadataKey] = normalizedVisitorId;
        }
        return ReadOnlyMetadata(merged);
    }

    private static string? ValidVisitorId(string? visitorId)
    {
        var normalizedVisitorId = visitorId?.Trim();
        if (string.IsNullOrEmpty(normalizedVisitorId) || normalizedVisitorId.Length > 128)
        {
            return null;
        }
        foreach (var character in normalizedVisitorId)
        {
            if (!char.IsAsciiLetterOrDigit(character) && character is not '_' and not '-')
            {
                return null;
            }
        }
        return normalizedVisitorId;
    }

    private static ReadOnlyDictionary<string, string> EmptyMetadata()
    {
        return ReadOnlyMetadata(new Dictionary<string, string>());
    }

    private static ReadOnlyDictionary<string, string> ReadOnlyMetadata(Dictionary<string, string> metadata)
    {
        return new ReadOnlyDictionary<string, string>(metadata);
    }
}

File: java/cekat-event-sdk-core/src/main/java/ai/cekat/events/stripe/StripeMetadata.java
package ai.cekat.events.stripe;

import ai.cekat.events.VisitorContext;
import java.util.LinkedHashMap;
import java.util.Map;

public final class StripeMetadata {
    private static final String VISITOR_METADATA_KEY = "cekat_visitor_id";

    private StripeMetadata() {
    }

    public static Map<String, String> forVisitor(String visitorId) {
        String normalized = validVisitorId(visitorId);
        return normalized == null ? Map.of() : Map.of(VISITOR_METADATA_KEY, normalized);
    }

    public static Map<String, String> fromCurrentVisitor() {
        return forVisitor(VisitorContext.currentVisitorId().orElse(null));
    }

    public static Map<String, String> mergeMetadata(Map<String, String> metadata, String visitorId) {
        LinkedHashMap<String, String> merged = new LinkedHashMap<>(metadata);
        String normalized = validVisitorId(visitorId);
        if (normalized != null) {
            merged.put(VISITOR_METADATA_KEY, normalized);
        }
        return Map.copyOf(merged);
    }

    private static String validVisitorId(String visitorId) {
        if (visitorId == null) {
            return null;
        }
        String normalized = visitorId.strip();
        if (normalized.isEmpty() || normalized.length() > 128) {
            return null;
        }
        for (int index = 0; index < normalized.length(); index++) {
            char character = normalized.charAt(index);
            if (!(character >= 'a' && character <= 'z')
                    && !(character >= 'A' && character <= 'Z')
                    && !(character >= '0' && character <= '9')
                    && character != '_' && character != '-') {
                return null;
            }
        }
        return normalized;
    }
}
